search/dijkstra.go

```diff
@@ -0,0 +1,187 @@
+		indexOf: indexOf,
+
+		dist: make([]float64, len(nodes)),
+		next: make([]int, len(nodes)),
+	}
+	for i := range path.dist {
+		path.dist[i] = math.Inf(1)
+	}
+
+	// Dijkstra's algorithm here is implemented essentially as
+	// described in Function B.2 in figure 6 of UTCS Technical
+	// Report TR-07-54.
+	//
+	// http://www.cs.utexas.edu/ftp/techreports/tr07-54.pdf
+	Q := priorityQueue{{node: u, dist: 0}}
+	for Q.Len() != 0 {
+		mid := heap.Pop(&Q).(distanceNode)
+		k := path.indexOf[mid.node.ID()]
+		if mid.dist < path.dist[k] {
+			path.dist[k] = mid.dist
+		}
+		for _, v := range from(mid.node) {
+			j := path.indexOf[v.ID()]
+			w := weight(edgeTo(mid.node, v))
+			if w < 0 {
+				panic("dijkstra: negative edge weight")
+			}
+			joint := path.dist[k] + w
+			if joint < path.dist[j] {
+				heap.Push(&Q, distanceNode{node: v, dist: joint})
+				path.set(j, joint, k)
+			}
+		}
+	}
+
+	return path
+}
+
+// DijkstraAllPaths returns a shortest-path tree for shortest paths in the graph g.
+// If weight is nil and the graph does not implement graph.Coster, UniformCost is used.
+// DijkstraAllPaths will panic if g has a negative edge weight.
+func DijkstraAllPaths(g graph.Graph, weight graph.CostFunc) (paths ShortestPaths) {
+	var (
+		from   = g.Neighbors
+		edgeTo func(graph.Node, graph.Node) graph.Edge
+	)
+	switch g := g.(type) {
+	case graph.DirectedGraph:
+		from = g.Successors
+		edgeTo = g.EdgeTo
+	default:
+		edgeTo = g.EdgeBetween
+	}
+	if weight == nil {
+		if g, ok := g.(graph.Coster); ok {
+			weight = g.Cost
+		} else {
+			weight = UniformCost
+		}
+	}
+
+	nodes := g.NodeList()
+
+	indexOf := make(map[int]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	dist := make([]float64, len(nodes)*len(nodes))
+	for i := range dist {
+		dist[i] = math.Inf(1)
+	}
+	paths = ShortestPaths{
+		nodes:   nodes,
+		indexOf: indexOf,
+
+		dist:    mat64.NewDense(len(nodes), len(nodes), dist),
+		next:    make([][]int, len(nodes)*len(nodes)),
+		forward: false,
+	}
+
+	var Q priorityQueue
+	for i, u := range nodes {
+		// Dijkstra's algorithm here is implemented essentially as
+		// described in Function B.2 in figure 6 of UTCS Technical
+		// Report TR-07-54 with the addition of handling multiple
+		// paths.
+		//
+		// http://www.cs.utexas.edu/ftp/techreports/tr07-54.pdf
+
+		Q = Q[:0]
+		heap.Push(&Q, distanceNode{node: u, dist: 0})
+		for Q.Len() != 0 {
+			mid := heap.Pop(&Q).(distanceNode)
+			k := paths.indexOf[mid.node.ID()]
+			if mid.dist < paths.dist.At(i, k) {
+				paths.dist.Set(i, k, mid.dist)
+			}
+			for _, v := range from(mid.node) {
+				j := paths.indexOf[v.ID()]
+				w := weight(edgeTo(mid.node, v))
+				if w < 0 {
+					panic("dijkstra: negative edge weight")
+				}
+				joint := paths.dist.At(i, k) + w
+				if joint < paths.dist.At(i, j) {
+					heap.Push(&Q, distanceNode{node: v, dist: joint})
+					paths.set(i, j, joint, k)
+				} else if joint == paths.dist.At(i, j) {
+					paths.add(i, j, k)
```

Comment:
I am not convinced that this is correct - I have a suspicion that we may miss some paths - but if I conditionally push v if it has not been seen (we add a []bool seen to do this), I don't see any cases that I would have missed with the tests we have.